relay/config/dynamic_config.go

```diff
@@ -7,47 +7,106 @@ import (
+		err = yaml.Unmarshal(buf, &currentYAML)
 		if err != nil {
-			log.Errorf("Parsing dynamic config for bundle %s failed: %s.", bundle, err)
-			return retval
+			log.Errorf("Parsing dynamic config file %s failed: %s.", f, err)
+			return make(map[string]interface{})
 		}
-		for k := range retval {
-			if strings.HasPrefix(k, "COG_") || strings.HasPrefix(k, "RELAY_") {
-				delete(retval, k)
-				log.Infof("Deleted illegal key %s from dynamic config for bundle %s.", k, bundle)
-			}
+
+		retval = shallowMerge(retval, currentYAML)
+	}
+
+	for k := range retval {
+		if strings.HasPrefix(k, "COG_") || strings.HasPrefix(k, "RELAY_") {
+			delete(retval, k)
+			log.Infof("Deleted illegal key %s from dynamic config for bundle %s.", k, bundle)
 		}
 	}
+
 	return retval
 }
 
-func locateConfigFile(configRoot string, bundle string) string {
-	rootInfo, rootErr := os.Stat(configRoot)
-	if rootErr != nil || rootInfo.IsDir() == false {
-		log.Debugf("Dynamic configuration root dir not found.")
-		return ""
+// Given the config root, a bundle name, a Cog username, and a chat
+// room name, return a list of all the filenames to be consolidated,
+// in the order they should be layered.
+func allConfigFiles(configRoot string, bundle string, room string, user string) []string {
+	var configs []string
+
+	if path := locateBaseConfigFile(configRoot, bundle); path != "" {
+		configs = append(configs, path)
+	}
+
+	if path := locateRoomConfigFile(configRoot, bundle, room); path != "" {
+		configs = append(configs, path)
 	}
-	fullYamlPath := path.Join(configRoot, bundle, "config.yaml")
-	fullYmlPath := path.Join(configRoot, bundle, "config.yml")
+
+	if path := locateUserConfigFile(configRoot, bundle, user); path != "" {
+		configs = append(configs, path)
+	}
+
+	return configs
+}
+
+// Given the config root, and a bundle name, return the base
+// configuration file.
+func locateBaseConfigFile(configRoot string, bundle string) string {
+	return resolveYAMLFile(path.Join(configRoot, bundle, "config"));
+}
+
+// Given the config root, a bundle name, and a room name, return the
+// path to the configuration file for that room.
+func locateRoomConfigFile(configRoot string, bundle string, roomName string) string {
+	roomFile := fmt.Sprintf("room_%s", strings.ToLower(roomName))
+	return resolveYAMLFile(path.Join(configRoot, bundle, roomFile));
+}
+
+// Given the config root, a bundle name, and a user name, return the
+// path to the configuration file for that user.
+func locateUserConfigFile(configRoot string, bundle string, userName string) string {
+	userFile := fmt.Sprintf("user_%s", strings.ToLower(userName))
+	return resolveYAMLFile(path.Join(configRoot, bundle, userFile))
+}
+
+// Given a base file name, return the path to either the yaml or yml
+// version, returning the ".yaml" version preferentially.
+func resolveYAMLFile(base string) string {
+
+	fullYamlPath := fmt.Sprint(base, ".yaml")
+	fullYmlPath := fmt.Sprint(base, ".yml")
```

Comment:
Micro-optimization -- Move `fullYmlPath := fmt.Sprint(base, ".yml")` inside the `if` so we don't do the work if the `.yaml` file is found.